[WIP] `STRTree` parallel implementation

darribas opened the pull request:
This takes the single-core implementation in #110 and adds multi-core processing for the spatial index query and the intersection/area calculation steps. Very much work in progress so not to merge yet, but I wanted to get the logic down and start a PR. @martinfleis and are hopefully working on this in early Jan and shipping a fully fledged version soon, it'd be great if we could take this one over the line for the January release of the metapackage? At any rate, this will go as "experimental" (possibly in its own separate function, independent of `_area_tables_binning` to show it's bleeding edge.

cc' @martinfleis, @knaaptime, @sjsrey

knaaptime commented:
does this supersede #110? or should we merge that while this is wip?

martinfleis commented:
@knaaptime this builds on top of #110. We can merge that while this is in WIP.

sjsrey commented:
> @knaaptime this builds on top of #110. We can merge that while this is in WIP.

#110 has been merged.

darribas commented:
OK this implementation is ready and a test is added. this may fail as the parallel implementation requires `master` versions of `pygeos` and `geopandas`.

martinfleis commented on tobler/area_weighted/area_interpolate.py:
```diff
@@ -9,10 +9,135 @@
+
+
+def _intersect_area_on_chunk(geoms1, geoms2):
+    import pygeos
+
+    intersection = pygeos.intersection(geoms1, geoms2)
+    areas = pygeos.measurement.area(intersection)
```
```suggestion
    intersection = geoms1.intersection(geoms2)
    areas = intersection.area
```

`geoms1` and `geoms1` are not Numpy arrays of pygeos geometries here, are they? If I am reading it correctly it is `GeometryArray` or `GeoSeries`, so the code above should work as well without a direct import of pygeos. It also makes it work with pure shapely without pygeos (not that one would want to do that).

darribas replied:
Great point!

martinfleis commented on tobler/area_weighted/area_interpolate.py:
```diff
@@ -9,10 +9,135 @@
+    # Spatial index query
+    ## Reindex on positional IDs
+    to_workers = _chunk_dfs(
+        gpd.GeoSeries(to_chunk.geometry.values, crs=to_chunk.crs),
+        gpd.GeoSeries(df_full.geometry.values, crs=df_full.crs),
```
Why are you creating new `GeoSeries` out of existing? `gpd.GeoSeries(to_chunk.geometry.values, crs=to_chunk.crs)` should be equal to `to_chunk.geometry` (the only difference could be `name`).

darribas replied:
We create those `GeoSeries` to use positional indices rather than `to_chunk.index` which is not positional and may have duplicates

martinfleis commented on tobler/tests/test_interpolators.py:
```diff
@@ -67,6 +67,19 @@ def test_area_interpolate_sindex_options():
         )
 
 
+def test_area_interpolate_parallel():
```
```suggestion
@pytest.mark.xfail(reason="requires master of pygeos and geopandas")
def test_area_interpolate_parallel():
```

knaaptime commented:
tests are passing so im happy to merge this and get the new release ready. Following up on your earlier comment, do you want to add a `warn` if someone invokes `n_jobs` just to note that its in beta. Since we have the notebook and its not a breaking change i'm not sure its absolutely critical... presumably it shouldn't be too long before the required version of pygeos is released anyway?

martinfleis commented:
GeoPandas 0.9 is expected within a month, pygeos 0.9 also soon (guessing from GitHub discussions).

Tests are passing because there's `xfail` decorator on the parallel one, which would otherwise fail.

darribas commented:
I think a warning is in order, will try to get to it asap :)

darribas commented:
OK @knaaptime this is good to go as far as Martin and I is concerned. I've moved the performance notebook to the notebooks folder and added a note on the documentation for area_interpolate that if `n_jobs` is not 1 you need edge pygeos/geopandas. Once that is released, we can remove the note or modify it with the version requirements.

Should we go ahead, merge and release so this makes it into the January PySAL release?
